internal/scheduler/nodedb.go

```diff
@@ -0,0 +1,332 @@
+package scheduler
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-memdb"
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/G-Research/armada/internal/scheduler/schedulerobjects"
+)
+
+// NodeDb is the scheduler-internal system for storing node information.
+// It's used to efficiently find nodes on which a pod can be scheduled.
+type NodeDb struct {
+	// In-memory database. Stores *SchedulerNode.
+	// Used to efficiently iterate over nodes in sorted order.
+	Db *memdb.MemDB
+	// Allowed pod priorities in sorted order.
+	// Because the number of database indices scales linearly with the number of distinct priorities,
+	// the efficiency of the NodeDb relies on the number of distinct priorities being small.
+	priorities       []int32
+	indexedResources map[string]interface{}
+	// Total amount of resources, e.g., "cpu", "memory", "gpu", managed by the scheduler.
+	// Computed approximately by periodically scanning all nodes in the db.
+	totalResources map[string]*resource.Quantity
+	// Set of node types for which there exists at least 1 node in the db.
+	NodeTypes map[string]*schedulerobjects.NodeType
```

Comment:
do we need to export all this outside the module?

Reply:
Not really. We can change it.